### DISCUSSION TITLE
'Timestamp' , 'Decimal' and 'datetime' is not JSON serializable

### DISCUSSION THREAD

**Discussion by command-z-z** [apache/airflow#30109] [General]
### Apache Airflow version

2.5.1

### What happened

Json can't encoder some datetype such as `Timestamp` `Decimal` `datetime` and so on.
But when i use Operator, which will print data of these datatype.
![557a590199437157a91052cb62d99244](https://user-images.githubusercontent.com/71259900/224880628-118a63df-723e-466d-8b3c-df1e73d27ee8.png)
![48872ca3a46d21f243333aa76461db75](https://user-images.githubusercontent.com/71259900/224880671-ea4f5fd1-331e-4187-b8ad-a6ac539c9fdc.png)

### How to reproduce

Simplified reproducer:
```python
import json
from decimal import Decimal
import pandas as pd
import numpy as np

dt = datetime.now() # get current timestamp
tp = pd.Timestamp("2023-03-30T12")
de = Decimal('3.14')

json_str = json.dumps({'test': dt})
print(json_str)
```

Some possible suggestions:
Airflow can judge the data type, maybe fix:
```python
if isinstance(data, Decimal):
    json_str = json.dumps({'test': dt}, default=str)
...
```

### Operating System

ALL Linux specifically ubuntu

### Versions of Apache Airflow Providers

not use any providers

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
Your reproducing example is not related to Airflow, it's better to provide an example similar to the real problem, but fortunately the first line in the screenshot can help to understand the problem.

You get this error when XComs try to serialize these fields which are not serializable.

> json_str = json.dumps({'test': dt}, default=str)

We cannot do that because we need to deserialize the values later in order to use them, and if we convert them to string, we lose all the information about the original types.
In the XComs [doc](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html#xcoms), we mention that you can use it with any **serializable** value, and since XComs uses json as default serializable, you cannot use XComs with them.

However, XComs supports pickle as serializer, so if you want to use it with no json serializable values, you can [enable XCom pickling](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#enable-xcom-pickling) (note that this is insecure and allows for RCE exploits)
